src/query/ee/src/storages/fuse/operations/vacuum_table.rs

```diff
@@ -13,329 +13,510 @@
+        );
+        // iterator snapshot files with time in descending order
+        for (i, (timestamp, snapshot_file)) in snapshot_files_with_time.iter().enumerate() {
+            let last_snapshot_location = if let Some(timestamp) = timestamp {
+                if timestamp >= retention_time {
+                    // save the last snapshot location that timestamp >= retention_time and continue the next loop
+                    last_snapshot_location_opt = Some(snapshot_file);
+                    continue;
+                } else if let Some(last_snapshot_location) = last_snapshot_location_opt {
+                    // now we get the first snapshot that timestamp < retention_time, return last_snapshot_location
+                    last_snapshot_location
+                } else {
+                    debug_assert!(i == 0);
+                    // when last_snapshot_location_opt == None, check if current snapshot is snapshot_files_with_time[0]
+                    let current_snapshot_location = match self.fuse_table.snapshot_loc().await? {
+                        Some(current_snapshot_location) => current_snapshot_location,
+                        None => {
+                            status = format!(
+                                "do_vacuum with table {}: read current snapshot location fail",
+                                self.fuse_table.get_table_info().name,
+                            );
+                            break;
+                        }
+                    };
+
+                    // if current snapshot == last snapshot, return
+                    if &current_snapshot_location == snapshot_file {
+                        return Ok(Some((0, current_snapshot_location)));
+                    }
+
+                    status = format!(
+                        "do_vacuum with table {}: last snapshot timestamp < retention_time and it isnot current snapshot",
+                        self.fuse_table.get_table_info().name,
+                    );
+                    break;
+                }
+            } else {
+                // if cannot find a snapshot with timestamp and timestamp < retention_time, return None
+                status = format!(
+                    "do_vacuum with table {}: cannot find a snapshot filename with timestamp",
+                    self.fuse_table.get_table_info().name,
+                );
+                break;
+            };
+
+            // here we get the first snapshot that:
+            //  1. timestamp < retention_time
+            //  2. it is not the current snapshot
+
+            // check if last_snapshot_location'prev match snapshot_file
+            // if not it means that last snapshot or snapshot_files_with_time[i]
+            // has not been committed success
+            // in both cases, return None
+            let last_snapshot = match self
+                .fuse_table
+                .read_table_snapshot_by_location(last_snapshot_location.to_string())
```

Comment:
Maybe we can use `SnapshotsIO::read_snapshot` directly?
https://github.com/datafuselabs/databend/blob/60004ed525cb2e122e0b5b07810c853ec5b0b617/src/query/storages/fuse/src/io/snapshots.rs#L78-L92